users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py

```diff
@@ -107,35 +109,42 @@ def sox(self, audio, sample_rate, random_state):
         return audio
 
     @staticmethod
-    def preemphasis(audio, sample_rate, random_state, factor):
+    def preemphasis(audio, sample_rate, random_state, preemphasis_args):
         import numpy as np
 
-        def preemphasis_numpy(waveform, coeff=0.97):
+        def preemphasis_numpy(waveform, coeff):
             waveform = np.copy(waveform)
             waveform[..., 1:] -= coeff * waveform[..., :-1]
             return waveform
 
-        if random_state.random() < factor.prob:
-            preemphasis_coefficient = random_state.random() * (factor.max - factor.min) + factor.min
-            audio = preemphasis_numpy(audio, coeff=preemphasis_coefficient)
-
-        return audio
+        if random_state.random() < preemphasis_args.get("prob", None):
+            preemphasis_coefficient = random_state.random() * (
+                preemphasis_args.get("maximum", None) - preemphasis_args.get("minimum", None)
+            ) + preemphasis_args.get("minimum", None)
+            return preemphasis_numpy(audio, coeff=preemphasis_coefficient)
+        else:
+            return preemphasis_numpy(audio, coeff=preemphasis_args.get("default_coeff", 0.97))
 
     @staticmethod
     def apply_codecs(audio, sample_rate, random_state, codecs):
+        import numpy as np
         for codec in codecs:
-            prob = codec.pop("prob", 1.0)
-            if random_state.random() < prob:
-                if codec.get("encoding") == "ULAW":
-                    # check if audio is in the right range for mu-law encoding
-                    if np.max(np.abs(audio)) > 1.0:
-                        raise ValueError("Audio must be in the range [-1, 1] for mu-law encoding.")
-                    # standard value for mu-law encoding
-                    mu = 255.0
-                    # mu-law encoding formula
-                    audio = np.sign(audio) * np.log1p(mu * np.abs(audio)) / np.log1p(mu)
+            if codec.get("encoding") == "ULAW":
+                prob = codec.pop("prob", 1.0)
```

Comment:
Can't you also here use a `PerturbationFactor`"

Reply:
See above

Reply:
Codec is a special case since it is a list. As long as I just use mu law, it would be possible to use PerturbationFactor but to make it possible to add others later, I would keep it that way.